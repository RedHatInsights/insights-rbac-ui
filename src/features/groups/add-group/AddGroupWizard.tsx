import React, { useMemo, useRef, useState } from 'react';
import { useDispatch } from 'react-redux';
import useChrome from '@redhat-cloud-services/frontend-components/useChrome';
import { useFlag } from '@unleash/proxy-client-react';
import { addNotification } from '@redhat-cloud-services/frontend-components-notifications/';
import FormRenderer from '@data-driven-forms/react-form-renderer/form-renderer';
import Pf4FormTemplate from '@data-driven-forms/pf4-component-mapper/form-template';
import componentMapper from '@data-driven-forms/pf4-component-mapper/component-mapper';
import WarningModal from '@patternfly/react-component-groups/dist/dynamic/WarningModal';
import { schemaBuilder } from './schema';
import { addGroup, addServiceAccountsToGroup } from '../../../redux/groups/actions';
import { SetName } from './components/stepName/SetName';
import { SetRoles } from './components/stepRoles/SetRoles';
import { SetUsers } from './components/stepUsers/SetUsers';
import SetServiceAccounts from './components/stepServiceAccounts/SetServiceAccounts';
import { SummaryContent } from './components/stepReview/SummaryContent';
import { AddGroupSuccess } from './AddGroupSuccess';
import useAppNavigate from '../../../hooks/useAppNavigate';
import paths from '../../../utilities/pathnames';
import { AddGroupWizardContext } from './add-group-wizard-context';

const FormTemplate = (props: React.ComponentProps<typeof Pf4FormTemplate>) => <Pf4FormTemplate {...props} showFormControls={false} />;

// Component mapper extension for wizard steps
const mapperExtension = {
  'set-name': SetName,
  'set-roles': SetRoles,
  'set-users': SetUsers,
  'set-service-accounts': SetServiceAccounts,
  'summary-content': SummaryContent,
  'add-group-success': AddGroupSuccess,
};

interface AddGroupWizardProps {
  // No direct props - component receives data from URL params and context
}

export const AddGroupWizard: React.FC<AddGroupWizardProps> = () => {
  const dispatch = useDispatch();
  const navigate = useAppNavigate();
  const chrome = useChrome();

  // Feature flags for wizard configuration
  const enableServiceAccounts =
    (chrome.isBeta() && useFlag('platform.rbac.group-service-accounts')) ||
    (!chrome.isBeta() && useFlag('platform.rbac.group-service-accounts.stable'));

  const enableWorkspaces = useFlag('platform.rbac.workspaces');

  // When workspaces is enabled, roles are disabled (per business logic)
  const enableRoles = !enableWorkspaces;

  const [cancelWarningVisible, setCancelWarningVisible] = useState<boolean>(false);
  const [submitting] = useState<boolean>(false);
  const [submittingError] = useState<string | null>(null);
  const [wizardError, setWizardError] = useState<boolean | undefined>(undefined);
  const [wizardContextValue, setWizardContextValue] = useState<{
    success: boolean;
    submitting: boolean;
    error: string | null;
    setHideForm: () => void;
    setWizardSuccess: (success: boolean) => void;
  }>({
    success: false,
    submitting: false,
    error: null,
    setHideForm: () => {},
    setWizardSuccess: () => {},
  });

  const container = useRef<HTMLDivElement>(null);

  interface FormData {
    'group-name': string;
    'group-description': string;
    'users-list'?: Array<{ label: string; username?: string; value?: string }>;
    'roles-list'?: Array<{ uuid: string; value?: string }>;
    'set-service-accounts'?: string[];
  }

  const onSubmit = async (formData: FormData) => {
    const {
      'group-name': name,
      'group-description': description,
      'users-list': users,
      'roles-list': roles,
      'set-service-accounts': serviceAccounts,
    } = formData;

    // Prepare all group data for single API call
    const groupData = {
      name,
      description,
      // Add users if any are selected
      user_list: users && users.length > 0 ? users.map((user) => ({ username: user.username || user.label || user.value || '' })) : [],
      // Add roles if any are selected
      roles_list: roles && roles.length > 0 ? roles.map((role) => role.uuid || role.value || '').filter(Boolean) : [],
    };

    try {
      // Create group with all data in one call
      const newGroupAction = dispatch(addGroup(groupData)) as { payload: Promise<any> };
      const newGroup = await newGroupAction.payload;

      // Check if group creation actually succeeded
      if (newGroup?.error) {
        throw new Error('Group creation returned error: ' + JSON.stringify(newGroup.error));
      }

      // Handle service accounts separately (if enabled)
      if (serviceAccounts && serviceAccounts.length > 0) {
        const serviceAccountObjects = serviceAccounts.map((sa) => ({ uuid: sa }));
        const serviceAccountAction = dispatch(addServiceAccountsToGroup(newGroup.uuid, serviceAccountObjects)) as { payload: Promise<unknown> };
        await serviceAccountAction.payload;
      }

      // Success! Show notification
      dispatch(
        addNotification({
          variant: 'success',
          title: 'Group created successfully',
          description: 'The group has been created and configured successfully.',
        }),
      );

      // Navigate to the new group or groups list (in separate try-catch to not fail group creation)
      try {
        if (newGroup?.uuid) {
          const pathname = `${paths['group-detail'].link.split('/').slice(0, -1).join('/')}/${newGroup.uuid}`;
          navigate(pathname);
        } else {
          console.warn('⚠️ No UUID returned, navigating to groups list');
          navigate('/groups');
        }
      } catch (navError) {
        console.error('❌ Navigation error (but group was created successfully):', navError);
        // Still try to go back to groups list
        navigate('/groups');
      }
    } catch (error) {
      console.error('❌ Error creating group:', error);
      dispatch(
        addNotification({
          variant: 'danger',
          title: 'Error creating group',
          description: 'There was an error creating the group. Please try again.',
        }),
      );
    } finally {
    }
  };

  const onCancelWarningConfirm = () => {
    setCancelWarningVisible(false);
    navigate('/groups');
  };

  const schema = schemaBuilder(container.current, enableServiceAccounts, enableRoles);

  const contextValue = useMemo(
    () => ({
      ...wizardContextValue,
      success: false,
      submitting,
      submittingGroup: false,
      submittingServiceAccounts: false,
      error: wizardError !== undefined ? wizardError : submittingError || undefined,
      setHideForm: () => {},
      setWizardSuccess: (success: boolean) => setWizardContextValue((prev) => ({ ...prev, success })),
      setWizardError,
    }),
    [wizardContextValue, wizardError, submitting, submittingError, setWizardError],
  );

  return (
    <AddGroupWizardContext.Provider value={contextValue}>
      <div ref={container}>
        {cancelWarningVisible && (
          <WarningModal
            isOpen={cancelWarningVisible}
            title="Exit group creation?"
            confirmButtonLabel="Exit"
            onClose={() => setCancelWarningVisible(false)}
            onConfirm={onCancelWarningConfirm}
          >
            {'All unsaved changes will be lost. Are you sure you want to exit?'}
          </WarningModal>
        )}
        <FormRenderer
          schema={schema}
          subscription={{ values: true }}
          FormTemplate={FormTemplate}
          componentMapper={{ ...componentMapper, ...mapperExtension }}
          onSubmit={onSubmit}
          onCancel={() => setCancelWarningVisible(true)}
        />
      </div>
    </AddGroupWizardContext.Provider>
  );
};

export default AddGroupWizard;
