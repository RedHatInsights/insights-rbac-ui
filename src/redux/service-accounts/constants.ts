export const NO_DATA = 'no-data';
export const LAST_PAGE = 'last-page';
export const RESULTS = 'results';
export const ERROR = 'error';
