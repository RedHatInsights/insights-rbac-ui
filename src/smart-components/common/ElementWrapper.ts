import React, { cloneElement } from 'react';
import { useOutletContext } from 'react-router-dom';

interface ElementWrapperProps {
  path?: string;
  children: React.ReactElement;
}

export const ElementWrapper: React.FC<ElementWrapperProps> = ({ children, path }) => {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  const componentProps: any = useOutletContext();
  return cloneElement(children, { ...componentProps, ...((path && componentProps?.[path]) || {}) });
};

export default ElementWrapper;
