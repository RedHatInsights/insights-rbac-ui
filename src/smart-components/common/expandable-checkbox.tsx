import React from 'react';
import { useFormApi } from '@data-driven-forms/react-form-renderer';
import useFieldApi, { UseFieldApiConfig } from '@data-driven-forms/react-form-renderer/use-field-api';
import { Checkbox, ExpandableSection, FormGroup, Radio, Tooltip } from '@patternfly/react-core';
import OutlinedQuestionCircleIcon from '@patternfly/react-icons/dist/dynamic/icons/outlined-question-circle-icon';

export type ExpandableCheckboxItemType = {
  title: string;
  description: string;
  name: string;
  options?: { title: string; description: string; name: string }[];
};

const ExpandableCheckbox: React.FC<UseFieldApiConfig> = (props) => {
  const formOptions = useFormApi();
  const { items, input } = useFieldApi(props);
  const values = formOptions.getState().values;
  return (
    <FormGroup>
      {items?.map((item: ExpandableCheckboxItemType, key: number) => (
        <Checkbox
          inputClassName="pf-v5-u-mt-xs"
          key={key}
          isChecked={!!values[input.name]?.[item.name]}
          onChange={(_e, value) => {
            const newVal = value && item.options ? item.options?.[0]?.name : value;
            input.onChange({
              ...values[input.name],
              [item.name]: newVal,
            });
          }}
          label={
            <ExpandableSection toggleText={item.title} isIndented>
              {item.description}
              {item.options && (
                <FormGroup role="radiogroup" fieldId={`${item.name}-radiogroup`}>
                  {item.options?.map((option, key) => (
                    <Radio
                      key={key}
                      name={`${item.name}-radio`}
                      isChecked={values[input.name]?.[item.name] === option.name}
                      onChange={() => {
                        input.onChange({
                          ...values[input.name],
                          [item.name]: option.name,
                        });
                      }}
                      label={
                        <React.Fragment>
                          {option.title}{' '}
                          {option.description && (
                            <Tooltip content={option.description}>
                              <OutlinedQuestionCircleIcon />
                            </Tooltip>
                          )}
                        </React.Fragment>
                      }
                      id={`${item.name}-option-${key}`}
                    />
                  ))}
                </FormGroup>
              )}
            </ExpandableSection>
          }
          id={`${item.name}-checkbox`}
          name={item.name}
        />
      ))}
    </FormGroup>
  );
};

export default ExpandableCheckbox;
