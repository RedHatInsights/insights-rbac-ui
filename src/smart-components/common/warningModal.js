import React from 'react';
import PropTypes from 'prop-types';
import { Modal, ModalVariant, Button } from '@patternfly/react-core';
import { ExclamationTriangleIcon } from '@patternfly/react-icons';
import './warningModal.scss';

export const WarningModal = ({ type, isOpen, onModalCancel, onConfirmCancel }) => (
  <Modal
    title={
      <span className="ins-c-wizard__cancel-warning-header">
        <ExclamationTriangleIcon size="md" className="ins-c-wizard__cancel-warning-header--icon" />
        Exit {type} creation?
      </span>
    }
    variant={ModalVariant.small}
    className="ins-c-wizard__cancel-warning"
    isOpen={isOpen}
    onClose={onModalCancel}
    actions={[
      <Button key="confirm" variant="primary" onClick={onConfirmCancel}>
        Exit
      </Button>,
      <Button key="cancel" variant="link" onClick={onModalCancel}>
        Stay
      </Button>,
    ]}
    isFooterLeftAligned
  >
    <span> All inputs will be discarded.</span>
  </Modal>
);

WarningModal.propTypes = {
  type: PropTypes.string,
  isOpen: PropTypes.bool,
  onModalCancel: PropTypes.func,
  onConfirmCancel: PropTypes.func,
};
