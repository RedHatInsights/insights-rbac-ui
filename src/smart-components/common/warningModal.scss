.ins-c-wizard__cancel-warning-header {
    display: flex;
    align-items: center;
    &--icon {
        fill: var(--pf-global--warning-color--100);
        margin-right: var(--pf-global--spacer--md);
    }
}
