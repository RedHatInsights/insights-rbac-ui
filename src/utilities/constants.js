export const RBAC_API_BASE = `/api/rbac/v1`;
