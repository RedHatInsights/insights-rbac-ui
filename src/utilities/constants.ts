export const RBAC_API_BASE = `/api/rbac/v1` as const;
export const RBAC_API_BASE_2 = `/api/rbac/v2` as const;
export const COST_API_BASE = `/api/cost-management/v1` as const;
export const INVENTORY_API_BASE = `/api/inventory/v1` as const;

export const DEFAULT_MUA_BUNDLE = 'rhel' as const;

export const DEFAULT_ACCESS_GROUP_ID = 'default-access' as const;
