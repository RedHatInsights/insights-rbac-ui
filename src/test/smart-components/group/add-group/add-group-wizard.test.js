import React from 'react';
import thunk from 'redux-thunk';
import { Provider } from 'react-redux';
import { shallow } from 'enzyme';
import configureStore from 'redux-mock-store';
import { shallowToJson } from 'enzyme-to-json';

import { MemoryRouter, Route } from 'react-router-dom';
import promiseMiddleware from 'redux-promise-middleware';
import { notificationsMiddleware } from '@redhat-cloud-services/frontend-components-notifications/';
import { RBAC_API_BASE } from '../../../../utilities/constants';
import AddGroupWizard from '../../../../smart-components/group/add-group/add-group-wizard';
import { ADD_NOTIFICATION } from '@redhat-cloud-services/frontend-components-notifications/index';
import { WarningModal } from '../../../../smart-components/common/warningModal';

describe('<AddGroupWizard />', () => {
  let initialProps;
  let initialState;
  const middlewares = [thunk, promiseMiddleware, notificationsMiddleware()];
  let mockStore;

  const GroupWrapper = ({ store, children }) => (
    <Provider store={store}>
      <MemoryRouter initialEntries={['/groups/', '/groups/add-group/', '/groups/']} initialIndex={1}>
        {children}
      </MemoryRouter>
    </Provider>
  );

  beforeEach(() => {
    initialProps = {
      uuid: '123',
    };

    initialState = {
      roleReducer: {
        roles: [
          {
            label: 'foo',
            value: 'bar',
          },
        ],
      },
      groupReducer: {
        groups: {
          data: [
            {
              uuid: '123',
              name: 'SampleGroup',
            },
          ],
        },
      },
      userReducer: {
        selectedUser: {},
        isUserDataLoading: false,
        users: [],
      },
    };
    mockStore = configureStore(middlewares);
  });

  it('should render correctly', () => {
    const store = mockStore(initialState);
    apiClientMock.get(`${RBAC_API_BASE}/groups`, mockOnce({ body: { data: [] } }));
    apiClientMock.get(`${RBAC_API_BASE}/roles`, mockOnce({ body: { data: [] } }));
    const wrapper = shallow(
      <GroupWrapper store={store}>
        <AddGroupWizard {...initialProps} />
      </GroupWrapper>
    ).dive();

    setImmediate(() => {
      expect(shallowToJson(wrapper)).toMatchSnapshot();
    });
  });

  it('should show a warning modal on Cancel', () => {
    const store = mockStore(initialState);

    apiClientMock.put(
      `${RBAC_API_BASE}/groups/`,
      mockOnce((req, res) => {
        expect(req).toBeTruthy();
        return res.status(200);
      })
    );

    apiClientMock.get(
      `${RBAC_API_BASE}/groups/`,
      mockOnce((req, res) => {
        expect(req).toBeTruthy();
        return res.status(200).body({ data: [] });
      })
    );

    apiClientMock.get(
      `${RBAC_API_BASE}/roles/`,
      mockOnce((req, res) => {
        expect(req).toBeTruthy();
        return res.status(200).body({ data: [] });
      })
    );

    const wrapper = mount(
      <GroupWrapper store={store}>
        <Route path="/groups/add-group/" render={() => <AddGroupWizard {...initialProps} />} />
      </GroupWrapper>
    );
    expect(wrapper.find(WarningModal).getElement().props.isOpen).not.toBeTruthy();
    const input = wrapper.find('input#group-name');
    input.getDOMNode().value = 'foo';
    input.simulate('change');
    wrapper.update();
    wrapper.find('.pf-m-link').simulate('click');
    wrapper.update();
    expect(wrapper.find(WarningModal).getElement().props.isOpen).toBeTruthy();
  });

  it('should not show a warning modal on Cancel when clean', (done) => {
    const store = mockStore(initialState);

    apiClientMock.put(
      `${RBAC_API_BASE}/groups/`,
      mockOnce((req, res) => {
        expect(req).toBeTruthy();
        return res.status(200);
      })
    );

    apiClientMock.get(
      `${RBAC_API_BASE}/groups/`,
      mockOnce((req, res) => {
        expect(req).toBeTruthy();
        return res.status(200).body({ data: [] });
      })
    );

    apiClientMock.get(
      `${RBAC_API_BASE}/roles/`,
      mockOnce((req, res) => {
        expect(req).toBeTruthy();
        return res.status(200).body({ data: [] });
      })
    );

    const wrapper = mount(
      <GroupWrapper store={store}>
        <Route path="/groups/add-group/" render={() => <AddGroupWizard {...initialProps} />} />
      </GroupWrapper>
    );
    const expectedActions = expect.arrayContaining([
      expect.objectContaining({
        type: ADD_NOTIFICATION,
        payload: expect.objectContaining({ title: 'Adding group', variant: 'warning' }),
      }),
    ]);
    expect(wrapper.find(WarningModal).getElement().props.isOpen).not.toBeTruthy();
    wrapper.find('.pf-m-link').simulate('click');

    setImmediate(() => {
      expect(store.getActions()).toEqual(expectedActions);
      expect(wrapper.find(WarningModal)).toHaveLength(0);
      done();
    });
  });
});
