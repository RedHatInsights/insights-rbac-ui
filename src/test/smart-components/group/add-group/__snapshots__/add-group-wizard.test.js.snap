// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AddGroupWizard /> should not show a warning modal on Cancel when clean 1`] = `
<ContextProvider
  value={
    Object {
      "store": Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "subscription": Subscription {
        "handleChangeWrapper": [Function],
        "listeners": Object {
          "notify": [Function],
        },
        "onStateChange": [Function],
        "parentSub": undefined,
        "store": Object {
          "clearActions": [Function],
          "dispatch": [Function],
          "getActions": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        },
        "unsubscribe": null,
      },
    }
  }
>
  <MemoryRouter
    initialEntries={
      Array [
        "/groups/",
        "/groups/add-group/",
        "/groups/",
      ]
    }
    initialIndex={1}
  >
    <Connect(AddGroupWizard)
      uuid="123"
    />
  </MemoryRouter>
</ContextProvider>
`;
